code/_onclick/hud/map_popups.dm

```diff
@@ -1,101 +1,165 @@
+ */
 /obj/screen
+	/**
+	 * Map name assigned to this object.
+	 * Automatically set by /client/proc/add_obj_to_map.
+	 */
 	var/assigned_map
-	var/list/screen_info = list()//x,x pix, y, y pix || x,y
-	var/del_on_map_removal = TRUE//this could probably be changed to be a proc, for conditional removal. for now, this works.
+	/**
+	 * Mark this object as garbage-collectible after you clean the map
+	 * it was registered on.
+	 *
+	 * This could probably be changed to be a proc, for conditional removal.
+	 * But for now, this works.
+	 */
+	var/del_on_map_removal = TRUE
 
-/client/proc/clear_map(var/map_to_clear)//not really needed most of the time, as the client's screen list gets reset on relog. any of the buttons are going to get caught by garbage collection anyway. they're effectively qdel'd.
-	if(!map_to_clear|| !(map_to_clear in screen_maps))
-		return FALSE
-	for(var/obj/screen/x in screen_maps[map_to_clear])
-		screen_maps[map_to_clear] -= x
-		if(x.del_on_map_removal)
-			qdel(x)
-	screen_maps -= map_to_clear
+/**
+ * A generic background object.
+ * It is also implicitly used to allocate a rectangle on the map, which will
+ * be used for auto-scaling the map.
+ */
+/obj/screen/background
+	name = "background"
+	icon = 'icons/mob/map_backgrounds.dmi'
+	icon_state = "clear"
+	layer = GAME_PLANE
+	plane = GAME_PLANE
 
-/client/proc/clear_all_maps()
-	for(var/x in screen_maps)
-		clear_map(x)
+/**
+ * Sets screen_loc of this screen object, in form of point coordinates,
+ * with optional pixel offset (px, py).
+ *
+ * If applicable, "assigned_map" has to be assigned before this proc call.
+ */
+/obj/screen/proc/set_position(x, y, px = 0, py = 0)
+	if(assigned_map)
+		screen_loc = "[assigned_map]:[x]:[px],[y]:[py]"
+	else
+		screen_loc = "[x]:[px],[y]:[py]"
 
-/client/proc/close_popup(var/popup)
-	winshow(src,popup,0)
-	handle_popup_close(popup)
+/**
+ * Sets screen_loc to fill a rectangular area of the map.
+ *
+ * If applicable, "assigned_map" has to be assigned before this proc call.
+ */
+/obj/screen/proc/fill_rect(x1, y1, x2, y2)
+	if(assigned_map)
+		screen_loc = "[assigned_map]:[x1],[y1] to [x2],[y2]"
+	else
+		screen_loc = "[x1],[y1] to [x2],[y2]"
 
-/client/verb/handle_popup_close(window_id as text) //when the popup closes in any way (player or proc call) it calls this.
-	set hidden = TRUE
-	clear_map("[window_id]_map")
+/**
+ * Registers screen obj with the client, which makes it visible on the
+ * assigned map, and becomes a part of the assigned map's lifecycle.
+ */
+/client/proc/register_map_obj(obj/screen/screen_obj)
+	if(!screen_obj.assigned_map)
+		CRASH("Can't register [screen_obj] without 'assigned_map' property.")
+	if(!screen_maps[screen_obj.assigned_map])
+		screen_maps[screen_obj.assigned_map] = list()
+	// NOTE: Possibly an expensive operation
+	var/list/screen_map = screen_maps[screen_obj.assigned_map]
+	if(!screen_map.Find(screen_obj))
+		screen_map += screen_obj
+	if(!screen.Find(screen_obj))
+		screen += screen_obj
```

Comment:
```
	if(!(screen_obj in screen_obj.assigned_map))
		screen_obj.assigned_map += screen_obj
	if(!(screen_obj in screen))
		screen += screen_obj
```
This is not a change request, it's just telling you that this syntax is also valid in case you prefer it. Else, discard it, both work the same.